<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Название-проекта:-Классификация-оттока-клиента" data-toc-modified-id="Название-проекта:-Классификация-оттока-клиента-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Название проекта: Классификация оттока клиента</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Цель-работы:" data-toc-modified-id="Цель-работы:-1.0.1"><span class="toc-item-num">1.0.1&nbsp;&nbsp;</span>Цель работы:</a></span></li><li><span><a href="#План-проекта:" data-toc-modified-id="План-проекта:-1.0.2"><span class="toc-item-num">1.0.2&nbsp;&nbsp;</span>План проекта:</a></span></li><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-1.0.3"><span class="toc-item-num">1.0.3&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li></ul></li><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Предобработка" data-toc-modified-id="Предобработка-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Предобработка</a></span></li><li><span><a href="#Выбросы" data-toc-modified-id="Выбросы-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Выбросы</a></span></li><li><span><a href="#Корреляция" data-toc-modified-id="Корреляция-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Корреляция</a></span></li><li><span><a href="#Изучим-остальные-столбцы" data-toc-modified-id="Изучим-остальные-столбцы-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Изучим остальные столбцы</a></span></li><li><span><a href="#Подготовка-данных-к-машинному-обучению" data-toc-modified-id="Подготовка-данных-к-машинному-обучению-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Подготовка данных к машинному обучению</a></span></li><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Решающее-дерево" data-toc-modified-id="Решающее-дерево-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Решающее дерево</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Взвешивание-классов" data-toc-modified-id="Взвешивание-классов-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Взвешивание классов</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Решающее-дерево" data-toc-modified-id="Решающее-дерево-4.1.2"><span class="toc-item-num">4.1.2&nbsp;&nbsp;</span>Решающее дерево</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-4.1.3"><span class="toc-item-num">4.1.3&nbsp;&nbsp;</span>Случайный лес</a></span></li></ul></li><li><span><a href="#Upsampling" data-toc-modified-id="Upsampling-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Upsampling</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Решающее-дерево" data-toc-modified-id="Решающее-дерево-4.2.2"><span class="toc-item-num">4.2.2&nbsp;&nbsp;</span>Решающее дерево</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-4.2.3"><span class="toc-item-num">4.2.3&nbsp;&nbsp;</span>Случайный лес</a></span></li></ul></li><li><span><a href="#Downsampling" data-toc-modified-id="Downsampling-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Downsampling</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Бонус-модель-градиентного-бустинга" data-toc-modified-id="Бонус-модель-градиентного-бустинга-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Бонус модель градиентного бустинга</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Вывод-по-проекту" data-toc-modified-id="Вывод-по-проекту-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Вывод по проекту</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

Попробую использовать Байсесовскую оптимизацию, заместо gridsearch. Поэтому установите следующую библиотеку:

In [1]:
!pip install hyperopt
!pip install catboost

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.6/98.6 MB 11.1 MB/s eta 0:00:00


# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

## Название проекта: Классификация оттока клиента
#### Цель работы:  
Обучить модель классификации, которая должна предсказывать откажется ли клиент от услуг банка или нет.

#### План проекта:
1) Подготовить данные к построению модели. <br/>
2) Проверить какая модель лучше подойдет для решения задачи. <br/>
3) Избавиться от дисбаланса классов. <br/>
4) Протестировать модель.

#### Вывод:
Находится в конце работы.

## Подготовка данных

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
import warnings
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score
import plotly.express as px
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix
import plotly.figure_factory as ff
from sklearn.metrics import roc_curve 
from sklearn.metrics import roc_auc_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score
import plotly.graph_objs as go
from hyperopt import hp, fmin, tpe, STATUS_OK, Trials
from sklearn.metrics import classification_report
from sklearn.utils import shuffle
from sklearn.ensemble import GradientBoostingClassifier
from catboost import CatBoostClassifier
from sklearn.dummy import DummyClassifier
from sklearn.preprocessing import OneHotEncoder
from sklearn.impute import KNNImputer

warnings.filterwarnings('ignore')

In [6]:
try:
    df = pd.read_csv('/content/Churn.csv')
except:
    df = pd.read_csv('/datasets/Churn.csv')

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


В нашем распоряжении датасет, состоящий из различных признаков таких как кредитный рейтинг, сколько лет клиент является клиентом банка и другие. Целевым признаком для нас является показатель Exited, который показывает отказался ли клиент от услуг банка или нет. Рассмотрим данные по-подробнее.

Критечески важно проверить пропуски в данных, как можно заметить ~1000 пропусков есть в столбце Tenure (сколько лет клиент является клиентом), посмотрим на столбец:

### Предобработка

In [8]:
df[df['Tenure'].isna()]

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
30           31    15589475    Azikiwe          591     Spain  Female   39   
48           49    15766205        Yin          550   Germany    Male   38   
51           52    15768193  Trevisani          585   Germany    Male   36   
53           54    15702298   Parkhill          655   Germany    Male   41   
60           61    15651280     Hunter          742   Germany    Male   35   
...         ...         ...        ...          ...       ...     ...  ...   
9944       9945    15703923    Cameron          744   Germany    Male   41   
9956       9957    15707861      Nucci          520    France  Female   46   
9964       9965    15642785    Douglas          479    France    Male   34   
9985       9986    15586914     Nepean          659    France    Male   36   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
30       NaN       0.00              3          1               0   
48       NaN  103391.38              1          0               1   
51       NaN  146050.97              2          0               0   
53       NaN  125561.97              1          0               0   
60       NaN  136857.00              1          0               0   
...      ...        ...            ...        ...             ...   
9944     NaN  190409.34              2          1               1   
9956     NaN   85216.61              1          1               0   
9964     NaN  117593.48              2          0               0   
9985     NaN  123841.49              2          1               0   
9999     NaN  130142.79              1          1               0   

      EstimatedSalary  Exited  
30          140469.38       1  
48           90878.13       0  
51           86424.57       0  
53          164040.94       1  
60           84509.57       0  
...               ...     ...  
9944        138361.48       0  
9956        117369.52       1  
9964        113308.29       0  
9985         96833.00       0  
9999         38190.78       0  

[909 rows x 14 columns]

Посмею предположить, что показатель того, сколько лет человек является клиентом банка очень важен для нашей задачи. Поэтому избавимся от строк с людьми, у которых не указан этот показатель. Да мы потеряем 1/10 данных, но наши данные будут кристально чисты.

In [9]:
df = df.dropna()

In [10]:
df[df['Tenure'].isna()]

Empty DataFrame
Columns: [RowNumber, CustomerId, Surname, CreditScore, Geography, Gender, Age, Tenure, Balance, NumOfProducts, HasCrCard, IsActiveMember, EstimatedSalary, Exited]
Index: []

In [11]:
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9091 entries, 0 to 9998
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        9091 non-null   int64  
 1   CustomerId       9091 non-null   int64  
 2   Surname          9091 non-null   object 
 3   CreditScore      9091 non-null   int64  
 4   Geography        9091 non-null   object 
 5   Gender           9091 non-null   object 
 6   Age              9091 non-null   int64  
 7   Tenure           9091 non-null   float64
 8   Balance          9091 non-null   float64
 9   NumOfProducts    9091 non-null   int64  
 10  HasCrCard        9091 non-null   int64  
 11  IsActiveMember   9091 non-null   int64  
 12  EstimatedSalary  9091 non-null   float64
 13  Exited           9091 non-null   int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.0+ MB


In [13]:
df['Tenure'].unique()

array([ 2.,  1.,  8.,  7.,  4.,  6.,  3., 10.,  5.,  9.,  0.])

Вряд ли количество лет должно быть дробным, изменим это:

In [14]:
df['Tenure'] = df['Tenure'].astype('int64')

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9091 entries, 0 to 9998
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        9091 non-null   int64  
 1   CustomerId       9091 non-null   int64  
 2   Surname          9091 non-null   object 
 3   CreditScore      9091 non-null   int64  
 4   Geography        9091 non-null   object 
 5   Gender           9091 non-null   object 
 6   Age              9091 non-null   int64  
 7   Tenure           9091 non-null   int64  
 8   Balance          9091 non-null   float64
 9   NumOfProducts    9091 non-null   int64  
 10  HasCrCard        9091 non-null   int64  
 11  IsActiveMember   9091 non-null   int64  
 12  EstimatedSalary  9091 non-null   float64
 13  Exited           9091 non-null   int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.0+ MB


### Выбросы

Теперь проверим данные на выбросы:

In [16]:
df['Age'].describe()

count    9091.000000
mean       38.949181
std        10.555581
min        18.000000
25%        32.000000
50%        37.000000
75%        44.000000
max        92.000000
Name: Age, dtype: float64

In [17]:
fig = px.histogram(df, x='Age', histfunc='count')

fig.update_layout(
    title='График по количеству лет среди клиентов',
    xaxis_title='Возраст',
    yaxis_title='Количество пользователей',
)

fig.show()

Средний возраст ~39 лет, а медианный 37, выбросов нет. Проверим другие признаки:

In [18]:
df['Balance'].describe()

count      9091.000000
mean      76522.740015
std       62329.528576
min           0.000000
25%           0.000000
50%       97318.250000
75%      127561.890000
max      250898.090000
Name: Balance, dtype: float64

In [19]:
fig = px.histogram(df, x='Balance', histfunc='count', color_discrete_sequence=['purple'])

fig.update_layout(
    title='График по балансу на счету клиента',
    xaxis_title='Сумма',
    yaxis_title='Количество пользователей',
)

fig.show()

Средний баланс 76000 долларов или евро, что в принципе не важно, максимальный 250000, интересно, что 25% клиентов на счету имеют 0.

In [20]:
df['EstimatedSalary'].describe()

count      9091.000000
mean     100181.214924
std       57624.755647
min          11.580000
25%       51227.745000
50%      100240.200000
75%      149567.210000
max      199992.480000
Name: EstimatedSalary, dtype: float64

In [21]:
fig = px.histogram(df, x='EstimatedSalary', histfunc='count', color_discrete_sequence=['orange'])

fig.update_layout(
    title='График по предположительной зарплате клиента',
    xaxis_title='Сумма',
    yaxis_title='Количество пользователей',
)

fig.show()

Здесь тоже выбросов не обнаружено. Зато теперь знаем среднюю зарплату в год это примерно 100000, ну и медиана примерно такая же.

In [22]:
df['CreditScore'].describe()

count    9091.000000
mean      650.736553
std        96.410471
min       350.000000
25%       584.000000
50%       652.000000
75%       717.000000
max       850.000000
Name: CreditScore, dtype: float64

In [23]:
fig = px.histogram(df, x='CreditScore', histfunc='count', color_discrete_sequence=['grey'])

fig.update_layout(
    title='График по предположительной зарплате клиента',
    xaxis_title='Сумма',
    yaxis_title='Количество пользователей',
)

fig.show()

Интересно, минимальный показатель 350, а максимальный 850, почти в 2.5 раза больше. Средний 650, что видно по графику.

In [24]:
df['Tenure'].describe()

count    9091.000000
mean        4.997690
std         2.894723
min         0.000000
25%         2.000000
50%         5.000000
75%         7.000000
max        10.000000
Name: Tenure, dtype: float64

In [25]:
fig = px.histogram(df, x='Tenure', histfunc='count', color_discrete_sequence=['red'])

fig.update_layout(
    title='График по времени пользования банком (в годах)',
    xaxis_title='Сумма',
    yaxis_title='Количество пользователей',
    bargap=0.2,
    bargroupgap=0.1,
    xaxis=dict(
        tickmode="linear",
        dtick=1)
)

fig.show()

Видно банк довольно молодой, раз самыми старыми клиентами являются люди, которые пользуются им всего лишь 10 лет. (таких людей 446)

### Корреляция

In [27]:
corr_matrix = df.corr()

fig = go.Figure(data=go.Heatmap(
        z=corr_matrix.values,
        x=list(corr_matrix.columns),
        y=list(corr_matrix.index),
        colorscale='magma'
    ))


fig.show()

Прослеживается слабая корреляция между ушел ли клиент и его возрастом. Ну и совсем слабая между балансом и ушел ли клиент. Между количеством продуктов и балансом явно корреляция отрицательная.

Чтобы было проще анализировать, избавимся от столбца Surname, так как эта информация не влияет ни на что.

In [28]:
df.drop('Surname', inplace=True, axis=1)

In [29]:
df.columns

Index(['RowNumber', 'CustomerId', 'CreditScore', 'Geography', 'Gender', 'Age',
       'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

Изучим столбцы, которые мы еще не рассматривали, а для этого сделаем все столбцы более читаемыми, также избавимся от двух столбцов, которые будут только мешать нашим алгоритмам машинного обучения.

In [30]:
df.columns = ['row_number', 'customer_id', 'credit_score', 'geography', 'gender', 'age',
       'tenure', 'balance', 'num_of_products', 'has_credit_card', 'is_active_member',
       'estimated_salary', 'exited']
df = df.drop(['row_number', 'customer_id'], axis=1)

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9091 entries, 0 to 9998
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   credit_score      9091 non-null   int64  
 1   geography         9091 non-null   object 
 2   gender            9091 non-null   object 
 3   age               9091 non-null   int64  
 4   tenure            9091 non-null   int64  
 5   balance           9091 non-null   float64
 6   num_of_products   9091 non-null   int64  
 7   has_credit_card   9091 non-null   int64  
 8   is_active_member  9091 non-null   int64  
 9   estimated_salary  9091 non-null   float64
 10  exited            9091 non-null   int64  
dtypes: float64(2), int64(7), object(2)
memory usage: 852.3+ KB


### Изучим остальные столбцы

In [32]:
others = ['num_of_products', 'gender', 'has_credit_card', 'is_active_member', 'exited']

In [33]:
fig = px.histogram(df, x='num_of_products', color='num_of_products');
fig.update_layout(
    title_text="Столбчатая диаграмма по количеству продуктов, использованных пользователями",
    xaxis_title="Количество продуктов",
    yaxis_title="Количество пользователей",
    bargap=0.2,
    bargroupgap=0.1,
    xaxis=dict(
        tickmode="linear",
        dtick=1)
)
fig.show()

In [34]:
fig = px.histogram(df, x='gender', color='gender')

fig.update_layout(
    title_text="Столбчатая диаграмма по полу клиентов",
    xaxis_title="Пол",
    yaxis_title="Количество пользователей",
    bargap=0.2,
    bargroupgap=0.1,
    xaxis=dict(
        ticktext=["Мужской", "Женский"],
        tickvals=["Male", "Female"]
    ),
)

fig.show()


Как видно среди клиентов банка мужчин больше чем женщин. Хоть и всего на ~800 человек больше.

In [35]:
fig = px.histogram(df, x='has_credit_card', color='has_credit_card')

fig.update_layout(
    title_text="Столбчатая диаграмма по наличию кредитной карты у клиента",
    xaxis_title="Наличие",
    yaxis_title="Количество пользователей",
    bargap=0.2,
    bargroupgap=0.1,
    xaxis=dict(
        ticktext=["Не имеет", "Имеет"],
        tickvals=["0", "1"]
    ),
)

fig.show()


Как видно в банке преобладают люди с кредитными картами, причем в >2 раза.

In [36]:
fig = px.histogram(df, x='is_active_member', color='is_active_member')

fig.update_layout(
    title_text="Столбчатая диаграмма активен ли клиент",
    xaxis_title="Активен",
    yaxis_title="Количество пользователей",
    bargap=0.2,
    bargroupgap=0.1,
    xaxis=dict(
        ticktext=["Нет", "Да"],
        tickvals=["0", "1"]
    ),
)

fig.show()


Ситуация мягко говоря не очень, разница между количеством активных и количеством неактивных очень маленькая всего ~200 человек, хоть и преобладает категория активных клиентов, но все равно разница заставляет задуматься.

In [37]:
fig = px.histogram(df, x='exited', color='exited')

fig.update_layout(
    title_text="Столбчатая диаграмма активен ли клиент",
    xaxis_title="Ушел ли клиент",
    yaxis_title="Количество пользователей",
    bargap=0.2,
    bargroupgap=0.1,
    xaxis=dict(
        ticktext=["Нет", "Да"],
        tickvals=["0", "1"]
    ),
)

fig.show()


Интересно, около 1800 человек отказались от услуг банка, остальный 7200 все таки остаются клиентами, надолго ли? Именно эта задача стоит у нас, уйдет ли клиент или нет.

### Подготовка данных к машинному обучению

Как нам уже известно, для применения модели типа логистическая регрессия все признаки должны как минимум быть представлены числом, у нас есть такой столбец такой как geography, проверим какие в нем есть переменные.

In [38]:
df['geography'].unique()

array(['France', 'Spain', 'Germany'], dtype=object)

Всего три страны, можно использовать прямое кодирование.

In [39]:
df_test_valid, df_train = train_test_split(df, test_size=0.6, random_state=12345)
df_test, df_valid = train_test_split(df_test_valid, test_size=0.5, random_state=98765)
features_valid, features_test, features_train = df_valid.drop('exited', axis=1), df_test.drop('exited', axis=1), df_train.drop('exited', axis=1)
target_valid, target_test, target_train = df_valid['exited'], df_test['exited'], df_train['exited']

In [40]:
features_valid = pd.get_dummies(features_valid, drop_first=True)
features_test = pd.get_dummies(features_test, drop_first=True)
features_train = pd.get_dummies(features_train, drop_first=True)

In [41]:
features_train.head()

credit_score  age  tenure    balance  num_of_products  has_credit_card  \
862            725   41       7  113980.21                1                1   
9727           530   45       1       0.00                1                0   
1717           707   35       3   56674.48                1                1   
8640           730   32       9  127661.69                1                0   
5288           582   30       2       0.00                2                1   

      is_active_member  estimated_salary  geography_Germany  geography_Spain  \
862                  1         116704.25                  0                0   
9727                 1         190663.89                  0                0   
1717                 0          17987.40                  0                1   
8640                 0          60905.51                  0                0   
5288                 1         132029.95                  0                0   

      gender_Male  
862             1  
9727            0  
1717            0  
8640            0  
5288            1

Еще обязательно нужно масшабировать данные, так как возраст у нас от 18-92, баланс может достигать 250000, поэтому модель может подумать, что баланс важнее возраста, хотя возможно это не так, так что на всякий случай масштабируем данные.

In [42]:
scaler = StandardScaler()

In [43]:
scaler.fit(features_train)

StandardScaler()

In [44]:
features_train.iloc[:, :] = scaler.transform(features_train)
features_test.iloc[:, :] = scaler.transform(features_test)
features_valid.iloc[:, :] = scaler.transform(features_valid)

In [45]:
features_train.head()

credit_score       age    tenure   balance  num_of_products  \
862       0.753255  0.192290  0.687476  0.599531        -0.923939   
9727     -1.271203  0.569439 -1.392492 -1.219209        -0.923939   
1717      0.566382 -0.373432 -0.699169 -0.314875        -0.923939   
8640      0.805164 -0.656294  1.380799  0.817842        -0.923939   
5288     -0.731348 -0.844868 -1.045831 -1.219209         0.791543   

      has_credit_card  is_active_member  estimated_salary  geography_Germany  \
862          0.639969          0.961687          0.295639          -0.579396   
9727        -1.562575          0.961687          1.583206          -0.579396   
1717         0.639969         -1.039840         -1.422928          -0.579396   
8640        -1.562575         -1.039840         -0.675764          -0.579396   
5288         0.639969          0.961687          0.562444          -0.579396   

      geography_Spain  gender_Male  
862         -0.578268     0.918438  
9727        -0.578268    -1.088805  
1717         1.729303    -1.088805  
8640        -0.578268    -1.088805  
5288        -0.578268     0.918438

Кредитный рейтинг и баланс отрицательные, зато теперь они в одном масштабе!

Исследуем дисбаланс классов:

In [46]:
df['exited'][df['exited'] == 0].count()

7237

In [47]:
df['exited'][df['exited'] == 1].count()

1854

Определенно потребуется upsampling или downsampling, ну а пока подведем итоги.

### Вывод:
Мы подготовили данные к машинному обучению, а именно: убрали строки с пропусками в столбце Tenure, отказались от 1/10 строк, но получили кристально чистые данные, в которых мы уверены. Проверили данные на предмет аномалий, аномалии необнаружены, просмотрели каждый столбец, например узнали, что мужчины чаще пользуются услугами банка чем женщины, а также узнали, что большинство клиентов банка это обладатели кредитных карт. Избавились от столбца Surname, который будет только мешать при алгоритме машинного обучения случайный лес и решающее дерево, а также не будет позволять использовать модель логистической регрессии, так как тип столбца object. Использовали прямое кодирование для столбца geography, чтобы модель логистической регрессии работала, также изменили масштаб данных и поняли, что существует проблема с дисбалансом классов, которую дальше будем решать. Что же до задачи, то тут будем использовать обучение с учителем, проблема это типичная задача классификации, нужно определить уйдет ли клиент или нет. 

## Исследование задачи

Ориентироваться будем на показатель F1, поэтому для начала обучим логистическую регрессию, решающее дерево и случайный лес. Начнем с логистической регрессии.

### Логистическая регрессия

In [ ]:
model_lc = LogisticRegression(random_state=12345, solver='liblinear')
model_lc.fit(features_train, target_train)
pred_lc = model_lc.predict(features_valid)
probabilities_valid = model_lc.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

report = classification_report(target_valid, pred_lc)

In [ ]:
fpr, tpr, thresholds = roc_curve(target_valid, probabilities_one_valid)

trace_roc = go.Scatter(x=fpr, y=tpr, mode='lines', line=dict(color='blue'), name='ROC curve')

trace_diag = go.Scatter(x=[0, 1], y=[0, 1], mode='lines', line=dict(color='red'), name='Random constant model')

layout = go.Layout(
    title='ROC curve',
    xaxis=dict(title='False Positive Rate', range=[0, 1]),
    yaxis=dict(title='True Positive Rate', range=[0, 1])
)

fig = go.Figure(data=[trace_roc, trace_diag], layout=layout)

fig.show()


In [ ]:
print('AUC-SCORE:', roc_auc_score(target_valid, probabilities_one_valid))
print('F1-SCORE:', f1_score(target_valid, pred_lc))

AUC-SCORE: 0.7565692035963549
F1-SCORE: 0.2994011976047904


In [ ]:
print(report)

              precision    recall  f1-score   support

           0       0.83      0.96      0.89      1457
           1       0.54      0.21      0.30       361

    accuracy                           0.81      1818
   macro avg       0.68      0.58      0.59      1818
weighted avg       0.77      0.81      0.77      1818



Как видим для отрицательного класса полнота равна почти 0.96, что означает, модель почти правильно предсказывает отрицательный класс, а вот точность меньше, видимо модель больше ошибается в положительном классе, чем в отрицательном. Площадь ROC-кривой равна 0.75, что означает, что модель для предсказание положительного класса уже лучше константной. Построим матрицу ошибок:

In [ ]:

cm = confusion_matrix(target_valid, pred_lc)

fig = go.Figure(data=go.Heatmap(
        z=cm,
        x=['Предсказала 0', 'Предсказала 1'],
        y=['Правильно 0', 'Правильно 1'],
        colorscale='sunsetdark'
    ))

fig.update_layout(
    xaxis_title="Predicted",
    yaxis_title="True"
)

fig.show()


Как и было сказано, правильный предсказаний положительного класса всего 75, на 65 ошибок, что довольно плохо, из-за чего площадь под ROC-кривой такая маленькая, давайте попробуем сделать тюнинг гиперпараметров.

Теперь попробуем использовать Байесовский тюнинг гиперпараметров:

In [ ]:
space = {
    'penalty': hp.choice('penalty', ['l2']),
    'C': hp.choice('C', np.logspace(-5,5,20)),
    'solver': hp.choice('solver', ['newton-cg', 'sag', 'lbfgs', 'liblinear']),
    'max_iter': hp.choice('max_iter', range(50, 1001, 50))
}

def objective(params):
    clf = LogisticRegression(
        penalty=params['penalty'],
        C=params['C'],
        solver=params['solver'],
        max_iter=params['max_iter'],
        random_state=12345
    )

    clf.fit(features_train, target_train)

    y_pred = clf.predict(features_valid)

    score = f1_score(target_valid, y_pred)

    return {'loss': -score, 'params': params, 'status': STATUS_OK}

trials = Trials()
#зависит от версии numpy
try:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.RandomState(12345))
except:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.default_rng(12345))

    

print("Best hyperparameters:")
print(best)

best_model_clf = LogisticRegression(
    penalty='l2',
    C=np.logspace(-5,5,20)[best['C']],
    solver=['newton-cg', 'sag', 'lbfgs', 'liblinear'][best['solver']],
    max_iter=range(50, 1001, 50)[best['max_iter']],
    random_state=12345
)
best_model_clf.fit(features_train, target_train)

100%|██████████| 100/100 [00:41<00:00,  2.40trial/s, best loss: -0.3006012024048096]
Best hyperparameters:
{'C': 8, 'max_iter': 11, 'penalty': 0, 'solver': 1}


LogisticRegression(C=0.1623776739188721, max_iter=600, random_state=12345,
                   solver='sag')

In [ ]:

pred_clf = best_model_clf.predict(features_valid)

In [ ]:
probabilities_valid = best_model_clf.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
report = classification_report(target_valid, pred_clf)
print(report)

              precision    recall  f1-score   support

           0       0.83      0.96      0.89      1457
           1       0.54      0.21      0.30       361

    accuracy                           0.81      1818
   macro avg       0.69      0.58      0.59      1818
weighted avg       0.77      0.81      0.77      1818



In [ ]:
auc_lc = roc_auc_score(target_valid, probabilities_one_valid)
print('AUC-SCORE:', auc_lc)
print('F1-SCORE:', f1_score(target_valid, pred_clf))

AUC-SCORE: 0.7570996450415132
F1-SCORE: 0.3006012024048096


Да, только точность выросла на одну сотую и все. Площадь под ROC-кривой почти не изменилась, скорее всего это связано с тем, что даже после перебора гиперпараметров модель стала не сильно лучше предсказывать положительный класс.

In [ ]:
fpr, tpr, thresholds = roc_curve(target_valid, probabilities_one_valid)

trace_roc = go.Scatter(x=fpr, y=tpr, mode='lines', line=dict(color='blue'), name='ROC curve')

trace_diag = go.Scatter(x=[0, 1], y=[0, 1], mode='lines', line=dict(color='red'), name='Random constant model')

layout = go.Layout(
    title='ROC curve',
    xaxis=dict(title='False Positive Rate', range=[0, 1]),
    yaxis=dict(title='True Positive Rate', range=[0, 1])
)

fig = go.Figure(data=[trace_roc, trace_diag], layout=layout)

fig.show()


Проверим матрицу ошибок:

In [ ]:

cm = confusion_matrix(target_valid, pred_clf)

fig = go.Figure(data=go.Heatmap(
        z=cm,
        x=['Предсказала 0', 'Предсказала 1'],
        y=['Правильно 0', 'Правильно 1'],
        colorscale='sunsetdark'
    ))

fig.update_layout(
    xaxis_title="Predicted",
    yaxis_title="True"
)

fig.show()


Да некие улучшение произошли, но совсем незначительные, например она правильно определила еще 2 элемента отрицательного класса. Чтобы ускорить исследование будем сразу же подбирать гиперпараметры, чтобы получать "лучший результат".

### Решающее дерево

In [ ]:
space = {
    'criterion': hp.choice('criterion', ['gini', 'entropy']),
    'max_features': hp.choice('max_features', [None, 'sqrt', 'log2']),
    'max_depth': hp.choice('max_depth', range(2, 16)),
    'min_samples_leaf': hp.choice('min_samples_leaf', range(2, 16)),
    'min_samples_split': hp.choice('min_samples_split',range(3, 10))
}

def objective(params):
    clf = DecisionTreeClassifier(
        criterion=params['criterion'],
        max_depth=params['max_depth'],
        min_samples_split=params['min_samples_split'],
        min_samples_leaf=params['min_samples_leaf'],
        max_features=params['max_features'],
        random_state=12345
    )

    clf.fit(features_train, target_train)
    y_pred = clf.predict(features_valid)

    score = f1_score(target_valid, y_pred)

    return {'loss': -score, 'params': params, 'status': STATUS_OK}

trials = Trials()

try:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.RandomState(12345))
except:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.default_rng(12345))

print("Best hyperparameters:")
print(best)

best_model_dtc = DecisionTreeClassifier(
    max_depth=range(2, 16)[best['max_depth']],
    min_samples_split=range(3, 10)[best['min_samples_split']],
    min_samples_leaf=range(2, 16)[best['min_samples_leaf']],
    criterion=['gini', 'entropy'][best['criterion']],
    max_features=[None, 'sqrt', 'log2'][best['max_features']],
    random_state=12345
)
best_model_dtc.fit(features_train, target_train)
pred_dtc = best_model_dtc.predict(features_valid)
probabilities_valid = best_model_dtc.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]


100%|██████████| 100/100 [00:03<00:00, 32.84trial/s, best loss: -0.5662847790507365]
Best hyperparameters:
{'criterion': 1, 'max_depth': 5, 'max_features': 0, 'min_samples_leaf': 2, 'min_samples_split': 2}


In [ ]:
auc_dtc = roc_auc_score(target_valid, probabilities_one_valid)
print('AUC-SCORE:', auc_dtc)
print('F1-SCORE:', f1_score(target_valid, pred_dtc))

AUC-SCORE: 0.8183722862406531
F1-SCORE: 0.5662847790507365


In [ ]:
fpr, tpr, thresholds = roc_curve(target_valid, probabilities_one_valid)

trace_roc = go.Scatter(x=fpr, y=tpr, mode='lines', line=dict(color='blue'), name='ROC curve')

trace_diag = go.Scatter(x=[0, 1], y=[0, 1], mode='lines', line=dict(color='red'), name='Random constant model')

layout = go.Layout(
    title='ROC curve',
    xaxis=dict(title='False Positive Rate', range=[0, 1]),
    yaxis=dict(title='True Positive Rate', range=[0, 1])
)

fig = go.Figure(data=[trace_roc, trace_diag], layout=layout)

fig.show()


Мы добились показателя f1-меры в 0.57, когда как auc_roc стала 0.81, что тоже неплохо это и видно по графику. Теперь посмотрим на ConfusionMatrix, чтобы сделать окончательные выводы.

In [ ]:
pred_dtc = best_model_dtc.predict(features_valid)
cm = confusion_matrix(target_valid, pred_dtc)

fig = go.Figure(data=go.Heatmap(
        z=cm,
        x=['Предсказала 0', 'Предсказала 1'],
        y=['Правильно 0', 'Правильно 1'],
        colorscale='sunsetdark'
    ))

fig.update_layout(
    xaxis_title="Predicted",
    yaxis_title="True"
)

fig.show()


In [ ]:
report = classification_report(target_valid, pred_dtc)
print(report)

              precision    recall  f1-score   support

           0       0.88      0.95      0.91      1457
           1       0.69      0.48      0.57       361

    accuracy                           0.85      1818
   macro avg       0.79      0.71      0.74      1818
weighted avg       0.84      0.85      0.84      1818



Алгоритм машинного обучения решающее дерево дал результат порядком выше, чем логистическая регрессия, что значит, что мы двигаемся в правильном направлении. Площадь ROC кривой выросла на 7 сотых, а f1-мера стала 0.57, чего конечно недостаточно для решения задачи, но уже что-то посмотрим как изменилась полнота и точность. Precision и Recall тоже выросли, посмотрим подробнее:

In [ ]:
data = {'Модель': ['Логистическая регрессия', 'Решающее дерево'], 'Полнота': [recall_score(target_valid, pred_clf), recall_score(target_valid, pred_dtc)], 'Точность': [precision_score(target_valid, pred_clf), precision_score(target_valid, pred_dtc)]}
pd.DataFrame(data)

Модель   Полнота  Точность
0  Логистическая регрессия  0.207756  0.543478
1          Решающее дерево  0.479224  0.692000

Как видно точность выросла на 0.12, в то время как полнота всего на 0.29. Почему так произошло? Потому что в формуле точности в знаменателе фигурирует неправильные ответы положительного класса, в то время как они увеличились (62 -> 90), но это компенсировалось двухкратным увеличением количества правильных ответов. Почему так сильно выросла полнота? Потому что показатель неправильных  ответов отрицательного класса изменился (285 -> 178). AUC тоже выросла так как количество правильных ответов положительного класса увеличилось (76 -> 183), в то время как количество неправильных ответов положительного класса тоже увеличилось. Перейдем к последнему алгоритму машинного обучения, а именно случайный лес.

### Случайный лес

Как раз для этого нам и нужно было использовать Байесовский метод оптимизации гиперпараметров, мы в жизни не сможем сделать такой перебор GridSearch-м или HalvingGridSearch'м без большой потери времени. Предупреждение следующая ячейка выполнялась 4 минуты, поэтому в целях экономии времени, рекомендую пропускать.

In [49]:
%%time
space = {
    'criterion': hp.choice('criterion', ['gini', 'entropy']),
    'max_features': hp.choice('max_features', [None, 'sqrt', 'log2']),
    'max_depth': hp.choice('max_depth', range(2, 50)),
    'min_samples_leaf': hp.choice('min_samples_leaf', range(2, 35)), #
    'min_samples_split': hp.choice('min_samples_split',range(2, 35)), #
    'n_estimators': hp.choice('n_estimators',range(10, 210)) #
}

def objective(params):
    clf = RandomForestClassifier(
        criterion=params['criterion'],
        max_depth=params['max_depth'],
        min_samples_split=params['min_samples_split'],
        min_samples_leaf=params['min_samples_leaf'],
        max_features=params['max_features'],
        n_estimators = params['n_estimators'],
        random_state=12345
    )

    clf.fit(features_train, target_train)

    y_pred = clf.predict(features_valid)

    score = f1_score(target_valid, y_pred)

    return {'loss': -score, 'params': params, 'status': STATUS_OK}

trials = Trials()

try:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.RandomState(12345))
except:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.default_rng(12345))

print("Best hyperparameters:")
print(best)

best_model_rfc = RandomForestClassifier(
    max_depth=range(2, 50)[best['max_depth']],
    min_samples_split=range(2, 35)[best['min_samples_split']],
    min_samples_leaf=range(2, 35)[best['min_samples_leaf']],
    criterion=['gini', 'entropy'][best['criterion']],
    max_features=[None, 'sqrt', 'log2'][best['max_features']],
    n_estimators = range(10, 210)[best['n_estimators']],
    random_state=12345
)
best_model_rfc.fit(features_train, target_train)

100%|██████████| 100/100 [02:26<00:00,  1.46s/trial, best loss: -0.593698175787728]
Best hyperparameters:
{'criterion': 1, 'max_depth': 30, 'max_features': 0, 'min_samples_leaf': 1, 'min_samples_split': 20, 'n_estimators': 116}
CPU times: user 2min 21s, sys: 296 ms, total: 2min 21s
Wall time: 2min 28s


RandomForestClassifier(criterion='entropy', max_depth=32, max_features=None,
                       min_samples_leaf=3, min_samples_split=22,
                       n_estimators=126, random_state=12345)

Дублирую, чтобы не пришлось ждать:

In [50]:




best_model_rfc = RandomForestClassifier(criterion='entropy', max_depth=36, max_features=None,
                       min_samples_leaf=5, min_samples_split=7, n_estimators=40,
                       random_state=12345)
best_model_rfc.fit(features_train, target_train)

RandomForestClassifier(criterion='entropy', max_depth=32, max_features=None,
                       min_samples_leaf=3, min_samples_split=22,
                       n_estimators=126, random_state=12345)

In [ ]:
probabilities_valid = best_model_rfc.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
pred_rfc = best_model_rfc.predict(features_valid)
auc_rfc = roc_auc_score(target_valid, probabilities_one_valid)
print('AUC-SCORE:', auc_rfc)
print('F1-SCORE:', f1_score(target_valid, pred_rfc))

AUC-SCORE: 0.8417307220657937
F1-SCORE: 0.5983471074380166


In [ ]:
RandomForestClassifier(criterion='entropy', max_depth=32, max_features=None,
                       min_samples_leaf=3, min_samples_split=22,
                       n_estimators=126, random_state=12345)
best_model_rfc.fit(features_train, target_train)

In [51]:
probabilities_valid = best_model_rfc.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
pred_rfc = best_model_rfc.predict(features_valid)
auc_rfc = roc_auc_score(target_valid, probabilities_one_valid)
print('AUC-SCORE:', auc_rfc)
print('F1-SCORE:', f1_score(target_valid, pred_rfc))

AUC-SCORE: 0.8518889609241467
F1-SCORE: 0.593698175787728


Увеличили разброс перебора min_samples_leaf,min_samples_split, n_estimators и получили AUC-ROC выше, а F1 чуть ниже.

На валидационной выборке мы добились показателя в 0.59 f1-меры, площадь ROC-кривой 0.84. Это лучше чем было у Решающего Дерева, и решает нашу задачу однако: а) Данные несбалансированы, результат может быть лучше. б) Это на валидационной выборке, должно быть на тестовой, проверим тестовую:

In [ ]:
for_test = best_model_rfc.predict(features_test)
print('F1-SCORE:', f1_score(target_test, for_test))

F1-SCORE: 0.5648148148148149


In [52]:
for_test = best_model_rfc.predict(features_test)
print('F1-SCORE:', f1_score(target_test, for_test))

F1-SCORE: 0.550632911392405


Ожидаемо показатель упал на 0.03, почти проходим по критерию проекта, но все же недотягиваем. Посмотрим на матрицу ошибок.

In [53]:
cm = confusion_matrix(target_valid, pred_rfc)

fig = go.Figure(data=go.Heatmap(
        z=cm,
        x=['Предсказала 0', 'Предсказала 1'],
        y=['Правильно 0', 'Правильно 1'],
        colorscale='sunsetdark'
    ))

fig.update_layout(
    xaxis_title="Predicted",
    yaxis_title="True"
)

fig.show()


In [54]:
report = classification_report(target_valid, pred_rfc)
print(report)

              precision    recall  f1-score   support

           0       0.88      0.96      0.92      1457
           1       0.74      0.50      0.59       361

    accuracy                           0.87      1818
   macro avg       0.81      0.73      0.76      1818
weighted avg       0.86      0.87      0.85      1818



Ожидаемо, раз выросла f1-мера, но полнота, наоборот, уменьшалась, только точность выросла до 0.74. Это максимальные показатели, которые мы смогли получить. Давайте их сравним, но допишем еще парочку, чтобы было полное сравнение.

In [55]:
data = {'Модель': ['Логистическая регрессия', 'Решающее дерево', 'Случайный лес'], 'Полнота': [recall_score(target_valid, pred_clf), recall_score(target_valid, pred_dtc), recall_score(target_valid, pred_rfc)], 'Точность': [precision_score(target_valid, pred_clf), precision_score(target_valid, pred_dtc), precision_score(target_valid, pred_rfc)], 'F1-мера': [f1_score(target_valid, pred_clf), f1_score(target_valid, pred_dtc), f1_score(target_valid, pred_rfc)], 'AUC метрика': [auc_lc, auc_dtc, auc_rfc]}
pd.DataFrame(data)

NameError: ignored

In [ ]:
labels = ['Решающее дерево', 'Случайный лес', 'Логистическая регрессия']
auc = [auc_dtc, auc_rfc, auc_lc]
f1 = [f1_score(target_valid, pred_dtc), f1_score(target_valid, pred_rfc), f1_score(target_valid, pred_clf)]

fig = go.Figure(data=[
    go.Bar(name='AUC-ROC', x=labels, y=auc),
    go.Bar(name='F1-мера', x=labels, y=f1)
])

fig.show()


А тут уже интересно, полнота стала меньше, интересно почему же? В знаменателе формулы recall есть количество неправильных ответов отрицательного класса, а оно в решающем лесе было 178, а стало 180, то есть произошло ухудшение, но и количество правильных ответов положительного класса уменьшилось (183->181). То есть полнота законамерно уменьшилась, что неожиданно, но так как количество неправильных ответов положительного класса уменьшилось (90 -> 63), то точность возроcла на 0.07. А благодаря этому наша f1-мера увеличилась и стала ~0.6. Да мы просели по полноте, но по точности мы конпенсировали отставание. Наша модель случайного леса абсолютно точно стало лучше предсказывать правильный ответы отрицательного класса (1367->1394). Ну что же подведем итог:

### Вывод:
После подготовки данных мы приступили к решению задачи классификации путем обучения моделей машинного обучения без учета дисбаланса классов. Было проверено три алгоритма машинного обучения: Логистическая регрессия, Решающее дерево, Случайный лес. Все модели перед проверкой их f1-мер стали оптимальными, с помощью Байесовской оптимизации гиперпараметров мы получили лучший результат. В модели алгоритма логистической регрессии мы получили самый низкий результат 0.3, чуть по-лучше себя показал алгоритм решающего дерева, целых 0.57 по f1-мере просто отлично, ну а самым лучшим ожидаемо оказался алгоритм случайного леса, так как это ансамбль деревьев, то и показатели выше чем у решающего дерева. Целых 0.59, если бы и 0.59 было бы и на тестовой выборке, то мы бы решили задачу и никакой дисбаланс классов нам бы не помешал. По мере проверки f1-меры мы еще ориентировались на площадь под roc-кривой, здесь модели по местам расположились аналогично и у случайного леса целых 0.84, а минимальный у логистической регрессии целых 0.75, что кстати уже достаточно, так как мы знаем, что константная модель бы получила 0.5, то мы понимаем, что наша самая худшая модель по метрикам уже справляется со своей задачей. Перейдем к дисбалансу классов.

## Борьба с дисбалансом

Как будем бороться с дисбалансом классов? Можно проверить два способа, 1) Взвесить классы при обучении модели 2) Изменить размер выборки (upsampling, downsampling). Начнем с первого способа.

### Взвешивание классов

#### Логистическая регрессия

In [ ]:
space = {
    'penalty': hp.choice('penalty', ['l2']),
    'C': hp.choice('C', np.logspace(-5,5,20)),
    'solver': hp.choice('solver', ['newton-cg', 'sag', 'lbfgs', 'liblinear']),
    'max_iter': hp.choice('max_iter', range(50, 1001, 50)),
    'class_weight': hp.choice('class_weight', ['balanced'])
}

def objective(params):
    clf = LogisticRegression(
        penalty=params['penalty'],
        C=params['C'],
        solver=params['solver'],
        max_iter=params['max_iter'],
        class_weight=params['class_weight'],
        random_state=12345
    )

    clf.fit(features_train, target_train)

    y_pred = clf.predict(features_valid)

    score = f1_score(target_valid, y_pred)

    return {'loss': -score, 'params': params, 'status': STATUS_OK}

trials = Trials()
#зависит от версии numpy
try:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.RandomState(12345))
except:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.default_rng(12345))

    

print("Best hyperparameters:")
print(best)

balanced_clf = LogisticRegression(
    penalty='l2',
    C=np.logspace(-5,5,20)[best['C']],
    solver=['newton-cg', 'sag', 'lbfgs', 'liblinear'][best['solver']],
    max_iter=range(50, 1001, 50)[best['max_iter']],
    class_weight='balanced',
    random_state=12345
)
balanced_clf.fit(features_train, target_train)

100%|██████████| 100/100 [00:48<00:00,  2.05trial/s, best loss: -0.5020491803278688]
Best hyperparameters:
{'C': 6, 'class_weight': 0, 'max_iter': 14, 'penalty': 0, 'solver': 3}


LogisticRegression(C=0.01438449888287663, class_weight='balanced', max_iter=750,
                   random_state=12345, solver='liblinear')

Один лишь параметр сразу же увеличил показатель f1-меры с 0.3 до 0.5. Хм, интересно, как изменилась площадь под ROC-кривой.

In [ ]:
probabilities_valid = balanced_clf.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
pred_balanced_clf = balanced_clf.predict(features_valid)
fpr, tpr, thresholds = roc_curve(target_valid, probabilities_one_valid)

trace_roc = go.Scatter(x=fpr, y=tpr, mode='lines', line=dict(color='blue'), name='ROC curve')

trace_diag = go.Scatter(x=[0, 1], y=[0, 1], mode='lines', line=dict(color='red'), name='Random constant model')

layout = go.Layout(
    title='ROC curve',
    xaxis=dict(title='False Positive Rate', range=[0, 1]),
    yaxis=dict(title='True Positive Rate', range=[0, 1])
)

fig = go.Figure(data=[trace_roc, trace_diag], layout=layout)

fig.show()


In [ ]:

auc_balanced_clf = roc_auc_score(target_valid, probabilities_one_valid)
print('AUC-SCORE:', auc_balanced_clf)
print('F1-SCORE:', f1_score(target_valid, pred_balanced_clf))

AUC-SCORE: 0.7628565507617254
F1-SCORE: 0.5020491803278688


А вот площадь под ROC-кривой почти не поменялась, проверим матрицу ошибок тогда:

In [ ]:
cm = confusion_matrix(target_valid, pred_balanced_clf)

fig = go.Figure(data=go.Heatmap(
        z=cm,
        x=['Предсказала 0', 'Предсказала 1'],
        y=['Правильно 0', 'Правильно 1'],
        colorscale='sunsetdark'
    ))

fig.update_layout(
    xaxis_title="Predicted",
    yaxis_title="True"
)

fig.show()


То что площадь под ROC-кривой почти не поменялась объясняется тем, что хоть и количество правильных ответов положительного класса возросло, но и количество неправильных ответов тоже возросло, в итоге получилось так, что AUC-ROC не изменилась, а f1-мера еще как.

In [ ]:
report = classification_report(target_valid, pred_balanced_clf)
print(report)

              precision    recall  f1-score   support

           0       0.90      0.75      0.82      1457
           1       0.40      0.68      0.50       361

    accuracy                           0.73      1818
   macro avg       0.65      0.71      0.66      1818
weighted avg       0.80      0.73      0.75      1818



Сравним метрики до и после.

In [ ]:
labels = ['До', 'После']
auc = [precision_score(target_valid, pred_clf), precision_score(target_valid, pred_balanced_clf)]
f1 = [recall_score(target_valid, pred_clf), recall_score(target_valid, pred_balanced_clf)]

fig = go.Figure(data=[
    go.Bar(name='Точность', x=labels, y=auc),
    go.Bar(name='Полнота', x=labels, y=f1)
])

fig.show()


Точность упала, из-за того, что FPR стал больше, так как положительных предсказаний стало больше, полнота увеличилась, так как FNR стало меньше. Теперь посмотрим на другие модели.

#### Решающее дерево

In [ ]:
space = {
    'criterion': hp.choice('criterion', ['gini', 'entropy']),
    'max_features': hp.choice('max_features', [None, 'sqrt', 'log2']),
    'max_depth': hp.choice('max_depth', range(2, 16)),
    'min_samples_leaf': hp.choice('min_samples_leaf', range(2, 16)),
    'min_samples_split': hp.choice('min_samples_split',range(3, 10)),
    'class_weight': hp.choice('class_weight', ['balanced'])
}

def objective(params):
    clf = DecisionTreeClassifier(
        criterion=params['criterion'],
        max_depth=params['max_depth'],
        min_samples_split=params['min_samples_split'],
        min_samples_leaf=params['min_samples_leaf'],
        max_features=params['max_features'],
        class_weight=params['class_weight'],
        random_state=12345
    )

    clf.fit(features_train, target_train)
    y_pred = clf.predict(features_valid)

    score = f1_score(target_valid, y_pred)

    return {'loss': -score, 'params': params, 'status': STATUS_OK}

trials = Trials()

try:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.RandomState(12345))
except:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.default_rng(12345))

print("Best hyperparameters:")
print(best)

balanced_dtc = DecisionTreeClassifier(
    max_depth=range(2, 16)[best['max_depth']],
    min_samples_split=range(3, 10)[best['min_samples_split']],
    min_samples_leaf=range(2, 16)[best['min_samples_leaf']],
    criterion=['gini', 'entropy'][best['criterion']],
    max_features=[None, 'sqrt', 'log2'][best['max_features']],
    class_weight = 'balanced',
    random_state=12345
)
balanced_dtc.fit(features_train, target_train)
pred_balanced_dtc = balanced_dtc.predict(features_valid)
probabilities_valid = balanced_dtc.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]


100%|██████████| 100/100 [00:03<00:00, 31.63trial/s, best loss: -0.5608465608465609]
Best hyperparameters:
{'class_weight': 0, 'criterion': 1, 'max_depth': 6, 'max_features': 0, 'min_samples_leaf': 9, 'min_samples_split': 1}


Получили неожиданный результат, решающее дерево с учитыванием баланса классов справилась хуже чем без (отталкиваясь от f1-меры), странно, давайте проверим матрицу ошибок:

In [ ]:
cm = confusion_matrix(target_valid, pred_balanced_dtc)

fig = go.Figure(data=go.Heatmap(
        z=cm,
        x=['Предсказала 0', 'Предсказала 1'],
        y=['Правильно 0', 'Правильно 1'],
        colorscale='sunsetdark'
    ))

fig.update_layout(
    xaxis_title="Predicted",
    yaxis_title="True"
)

fig.show()


Взвешивание классов дало результат, но вот f1-мера не изменилась. Вероятнее всего это произошло из-за того, что модель стала просто больше предсказывать положительный класс, неправильных предсказаний не стало меньше, их число только увеличилось, перейдем к случайному лесу.

#### Случайный лес

Опять же рекомендую пропустить эту ячейку и сразу перейти к следующей, так как она долго выполняется:

In [56]:
%%time
space = {
    'criterion': hp.choice('criterion', ['gini', 'entropy']),
    'max_features': hp.choice('max_features', [None, 'sqrt', 'log2']),
    'max_depth': hp.choice('max_depth', range(2, 50)),
    'min_samples_leaf': hp.choice('min_samples_leaf', range(2, 35)),
    'min_samples_split': hp.choice('min_samples_split',range(2, 35)),
    'n_estimators': hp.choice('n_estimators',range(10, 210)),
    'class_weight': hp.choice('class_weight', ['balanced'])
}

def objective(params):
    clf = RandomForestClassifier(
        criterion=params['criterion'],
        max_depth=params['max_depth'],
        min_samples_split=params['min_samples_split'],
        min_samples_leaf=params['min_samples_leaf'],
        max_features=params['max_features'],
        n_estimators = params['n_estimators'],
        class_weight = params['class_weight'],
        random_state=12345
    )

    clf.fit(features_train, target_train)

    y_pred = clf.predict(features_valid)

    score = f1_score(target_valid, y_pred)

    return {'loss': -score, 'params': params, 'status': STATUS_OK}

trials = Trials()

try:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.RandomState(12345))
except:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.default_rng(12345))

print("Best hyperparameters:")
print(best)

balanced_rfc = RandomForestClassifier(
    max_depth=range(2, 50)[best['max_depth']],
    min_samples_split=range(2, 35)[best['min_samples_split']],
    min_samples_leaf=range(2, 35)[best['min_samples_leaf']],
    criterion=['gini', 'entropy'][best['criterion']],
    max_features=[None, 'sqrt', 'log2'][best['max_features']],
    n_estimators = range(10, 210)[best['n_estimators']],
    class_weight = 'balanced',
    random_state=12345
)
balanced_rfc.fit(features_train, target_train)

100%|██████████| 100/100 [01:55<00:00,  1.15s/trial, best loss: -0.6111111111111112]
Best hyperparameters:
{'class_weight': 0, 'criterion': 1, 'max_depth': 9, 'max_features': 1, 'min_samples_leaf': 2, 'min_samples_split': 29, 'n_estimators': 160}
CPU times: user 1min 43s, sys: 338 ms, total: 1min 43s
Wall time: 1min 56s


RandomForestClassifier(class_weight='balanced', criterion='entropy',
                       max_depth=11, min_samples_leaf=4, min_samples_split=31,
                       n_estimators=170, random_state=12345)

In [57]:
balanced_rfc = RandomForestClassifier(class_weight='balanced', criterion='entropy',
                       max_depth=11, min_samples_leaf=4, min_samples_split=31,
                       n_estimators=170, random_state=12345)
balanced_rfc.fit(features_train, target_train)

RandomForestClassifier(class_weight='balanced', criterion='entropy',
                       max_depth=11, min_samples_leaf=4, min_samples_split=31,
                       n_estimators=170, random_state=12345)

На удивление хоть и алгоритм решающего дерева подвел и дал показатель f1-мер хуже чем был, но случайный лес дал показатель на 0.015 больше чем при режиме без балансировки данных, давайте проверим как изменилась матрица ошибок.

In [58]:
pred_balanced_rfc = balanced_rfc.predict(features_valid)
cm = confusion_matrix(target_valid, pred_balanced_rfc)

fig = go.Figure(data=go.Heatmap(
        z=cm,
        x=['Предсказала 0', 'Предсказала 1'],
        y=['Правильно 0', 'Правильно 1'],
        colorscale='sunsetdark'
    ))

fig.update_layout(
    xaxis_title="Predicted",
    yaxis_title="True"
)

fig.show()


Число предсказаний положительного класса увеличилось, а значит число правильных ответов положительного класса тоже (181 -> 216). Число неправильных предсказаний положительного класса увеличилось в два раза (63 -> 125), число неправильных ответов отрицательного класса уменьшилось (180 -> 145), число правильных ответов отрицательного класса уменьшилось (1394 -> 1332). Посмотрим на показатель AUC-ROC:

In [ ]:
balanced_rfc = RandomForestClassifier(class_weight='balanced', max_depth=43,
                       max_features='sqrt', min_samples_leaf=4,
                       min_samples_split=9, n_estimators=67,
                       random_state=12345)
balanced_rfc.fit(features_train, target_train)

In [ ]:
probabilities_valid = balanced_rfc.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_rfc = roc_auc_score(target_valid, probabilities_one_valid)
print('AUC-SCORE:', auc_rfc)
print('F1-SCORE:', f1_score(target_valid, pred_balanced_rfc))

AUC-SCORE: 0.8495485544044702
F1-SCORE: 0.615819209039548


In [ ]:
balanced_rfc = RandomForestClassifier(class_weight='balanced', criterion='entropy',
                       max_depth=11, min_samples_leaf=4, min_samples_split=31,
                       n_estimators=170, random_state=12345)
balanced_rfc.fit(features_train, target_train)

In [59]:
probabilities_valid = balanced_rfc.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_rfc = roc_auc_score(target_valid, probabilities_one_valid)
print('AUC-SCORE:', auc_rfc)
print('F1-SCORE:', f1_score(target_valid, pred_balanced_rfc))

AUC-SCORE: 0.8574177197862263
F1-SCORE: 0.6111111111111112


Результат с увеличенным числом гиперпараметров лучше по AUC и хуже по f1

Раз число правильных ответов положительного класса увеличилось, то и площадь под ROC-кривой увеличилась, посмотрим на нее сейчас:

In [ ]:
fpr, tpr, thresholds = roc_curve(target_valid, probabilities_one_valid)

trace_roc = go.Scatter(x=fpr, y=tpr, mode='lines', line=dict(color='blue'), name='ROC curve')

trace_diag = go.Scatter(x=[0, 1], y=[0, 1], mode='lines', line=dict(color='red'), name='Random constant model')

layout = go.Layout(
    title='ROC curve',
    xaxis=dict(title='False Positive Rate', range=[0, 1]),
    yaxis=dict(title='True Positive Rate', range=[0, 1])
)

fig = go.Figure(data=[trace_roc, trace_diag], layout=layout)

fig.show()


Да, действительно, перед нами самая большая площадь под ROC-кривой за все время этого исследования. Проверим мало ли мы уже можем считать, что выполнили задачу и добились 0.59 f1-меры:

In [ ]:
test_pred = balanced_rfc.predict(features_test)
print('F1-SCORE:', f1_score(target_test, test_pred))

F1-SCORE: 0.5893089960886571


Все-таки чуть-чуть не хватило, видимо недостаточно просто взвесить классы, нужно менять выборку! Перейдем к этому шагу.

### Upsampling

Создадим функцию, которая сама увеличит выборку в нужное количество нам раз. На этапе подготовки данных мы поняли, что ответов отрицательного класса в нашем датасете примерно в 3.5 раза больше, чем ответов положительного класса. Увеличим нашу число ответов положительного класса в 4 раза и посмотрим, как это повлияет на f1-меру.

In [60]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled

In [61]:
features_upsampled, target_upsampled = upsample(features_train, target_train, 4)

#### Логистическая регрессия

In [ ]:
space = {
    'penalty': hp.choice('penalty', ['l2']),
    'C': hp.choice('C', np.logspace(-5,5,20)),
    'solver': hp.choice('solver', ['newton-cg', 'sag', 'lbfgs', 'liblinear']),
    'max_iter': hp.choice('max_iter', range(50, 1001, 50))
}

def objective(params):
    clf = LogisticRegression(
        penalty=params['penalty'],
        C=params['C'],
        solver=params['solver'],
        max_iter=params['max_iter'],
        random_state=12345
    )

    clf.fit(features_upsampled, target_upsampled)

    y_pred = clf.predict(features_valid)

    score = f1_score(target_valid, y_pred)

    return {'loss': -score, 'params': params, 'status': STATUS_OK}

trials = Trials()
#зависит от версии numpy
try:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.RandomState(12345))
except:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.default_rng(12345))

    

print("Best hyperparameters:")
print(best)

upsampled_clf = LogisticRegression(
    penalty='l2',
    C=np.logspace(-5,5,20)[best['C']],
    solver=['newton-cg', 'sag', 'lbfgs', 'liblinear'][best['solver']],
    max_iter=range(50, 1001, 50)[best['max_iter']],
    random_state=12345
)
upsampled_clf.fit(features_upsampled, target_upsampled)
pred_up_clf = upsampled_clf.predict(features_valid)
probabilities_valid = upsampled_clf.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

100%|██████████| 100/100 [00:51<00:00,  1.95trial/s, best loss: -0.498997995991984] 
Best hyperparameters:
{'C': 4, 'max_iter': 7, 'penalty': 0, 'solver': 0}


Метрика f1-мера стала чуть меньше (0.50 -> 0.49), чем при гиперпараметре class_weight=balanced, посмотрим на площадь под ROC-кривой.

In [ ]:
fpr, tpr, thresholds = roc_curve(target_valid, probabilities_one_valid)

trace_roc = go.Scatter(x=fpr, y=tpr, mode='lines', line=dict(color='blue'), name='ROC curve')

trace_diag = go.Scatter(x=[0, 1], y=[0, 1], mode='lines', line=dict(color='red'), name='Random constant model')

layout = go.Layout(
    title='ROC curve',
    xaxis=dict(title='False Positive Rate', range=[0, 1]),
    yaxis=dict(title='True Positive Rate', range=[0, 1])
)

fig = go.Figure(data=[trace_roc, trace_diag], layout=layout)

fig.show()


In [ ]:
print('AUC-ROC', roc_auc_score(target_valid, probabilities_one_valid))

AUC-ROC 0.7619154449719285


Площадь под ROC-кривой выросла на чуть-чуть по сравнению без балансировки классов. Теперь перейдем к решающему дереву:

#### Решающее дерево

In [ ]:
space = {
    'criterion': hp.choice('criterion', ['gini', 'entropy']),
    'max_features': hp.choice('max_features', [None, 'sqrt', 'log2']),
    'max_depth': hp.choice('max_depth', range(2, 16)),
    'min_samples_leaf': hp.choice('min_samples_leaf', range(2, 16)),
    'min_samples_split': hp.choice('min_samples_split',range(3, 10))
}

def objective(params):
    clf = DecisionTreeClassifier(
        criterion=params['criterion'],
        max_depth=params['max_depth'],
        min_samples_split=params['min_samples_split'],
        min_samples_leaf=params['min_samples_leaf'],
        max_features=params['max_features'],
        random_state=12345
    )

    clf.fit(features_train, target_train)
    y_pred = clf.predict(features_valid)

    score = f1_score(target_valid, y_pred)

    return {'loss': -score, 'params': params, 'status': STATUS_OK}

trials = Trials()

try:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.RandomState(12345))
except:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.default_rng(12345))

print("Best hyperparameters:")
print(best)

upsampled_dtc = DecisionTreeClassifier(
    max_depth=range(2, 16)[best['max_depth']],
    min_samples_split=range(3, 10)[best['min_samples_split']],
    min_samples_leaf=range(2, 16)[best['min_samples_leaf']],
    criterion=['gini', 'entropy'][best['criterion']],
    max_features=[None, 'sqrt', 'log2'][best['max_features']],
    random_state=12345
)
upsampled_dtc.fit(features_upsampled, target_upsampled)
pred_balanced_dtc = upsampled_dtc.predict(features_valid)
probabilities_valid = upsampled_dtc.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]


100%|██████████| 100/100 [00:02<00:00, 39.26trial/s, best loss: -0.5662847790507365]
Best hyperparameters:
{'criterion': 1, 'max_depth': 5, 'max_features': 0, 'min_samples_leaf': 2, 'min_samples_split': 2}


В целом все так же, как и без балансировки классов, как я уже сказал, ничего не поменяется по показателю f1-меры, давайте посмотрим на AUC-ROC:

In [ ]:
print('AUC-ROC:', roc_auc_score(target_valid, probabilities_one_valid))

AUC-ROC: 0.814005175131232


In [ ]:
fpr, tpr, thresholds = roc_curve(target_valid, probabilities_one_valid)

trace_roc = go.Scatter(x=fpr, y=tpr, mode='lines', line=dict(color='blue'), name='ROC curve')

trace_diag = go.Scatter(x=[0, 1], y=[0, 1], mode='lines', line=dict(color='red'), name='Random constant model')

layout = go.Layout(
    title='ROC curve',
    xaxis=dict(title='False Positive Rate', range=[0, 1]),
    yaxis=dict(title='True Positive Rate', range=[0, 1])
)

fig = go.Figure(data=[trace_roc, trace_diag], layout=layout)

fig.show()


Площадь под ROC-кривой чуть-чуть снизилась, f1-мера не поменялась. В целом очевидно, что ни логистическая регрессия, ни решающее дерево нам не пойдет для решения задачи, давайте перейдем к самому интересному, а именно к Случайному лесу.

#### Случайный лес

Следующую ячейку можно пропустить, чтобы не ждать.

In [62]:
%%time
space = {
    'criterion': hp.choice('criterion', ['gini', 'entropy']),
    'max_features': hp.choice('max_features', [None, 'sqrt', 'log2']),
    'max_depth': hp.choice('max_depth', range(2, 50)),
    'min_samples_leaf': hp.choice('min_samples_leaf', range(2, 35)),
    'min_samples_split': hp.choice('min_samples_split',range(2, 35)),
    'n_estimators': hp.choice('n_estimators',range(10, 210))
}

def objective(params):
    clf = RandomForestClassifier(
        criterion=params['criterion'],
        max_depth=params['max_depth'],
        min_samples_split=params['min_samples_split'],
        min_samples_leaf=params['min_samples_leaf'],
        max_features=params['max_features'],
        n_estimators = params['n_estimators'],
        random_state=12345
    )

    clf.fit(features_upsampled, target_upsampled)

    y_pred = clf.predict(features_valid)

    score = f1_score(target_valid, y_pred)

    return {'loss': -score, 'params': params, 'status': STATUS_OK}

trials = Trials()

try:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.RandomState(12345))
except:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.default_rng(12345))

print("Best hyperparameters:")
print(best)

upsampled_rfc = RandomForestClassifier(
    max_depth=range(2, 50)[best['max_depth']],
    min_samples_split=range(2, 35)[best['min_samples_split']],
    min_samples_leaf=range(2, 35)[best['min_samples_leaf']],
    criterion=['gini', 'entropy'][best['criterion']],
    max_features=[None, 'sqrt', 'log2'][best['max_features']],
    n_estimators = range(10, 210)[best['n_estimators']],
    random_state=12345
)
upsampled_rfc.fit(features_upsampled, target_upsampled)

100%|██████████| 100/100 [02:09<00:00,  1.30s/trial, best loss: -0.6]
Best hyperparameters:
{'criterion': 0, 'max_depth': 7, 'max_features': 1, 'min_samples_leaf': 18, 'min_samples_split': 30, 'n_estimators': 130}
CPU times: user 2min 5s, sys: 362 ms, total: 2min 6s
Wall time: 2min 10s


RandomForestClassifier(max_depth=9, min_samples_leaf=20, min_samples_split=32,
                       n_estimators=140, random_state=12345)

In [63]:
# upsampled_rfc = RandomForestClassifier(criterion='entropy', max_depth=16, max_features='sqrt',
#                       min_samples_leaf=11, min_samples_split=17,
#                       n_estimators=51, random_state=12345)

upsampled_rfc = RandomForestClassifier(max_depth=9, min_samples_leaf=20, min_samples_split=32,
                       n_estimators=140, random_state=12345)


upsampled_rfc.fit(features_upsampled, target_upsampled)
pred_balanced_rfc = upsampled_rfc.predict(features_valid)
probabilities_valid = upsampled_rfc.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

Здесь мы тоже добились показателя f1-меры в 0.6 на валидационной выборке, давайте проверим на тестовой.

In [64]:
test_pred = upsampled_rfc.predict(features_test)
print('F1-SCORE:', f1_score(target_test, test_pred))

F1-SCORE: 0.5933117583603021


Да на тестовой решение тоже проходит, что означает, что мы получили лучшую модель из всех, давайте посмотрим какой показатель AUC-ROC у лучшей модели.

In [65]:
print('AUC ROC:', roc_auc_score(target_valid, probabilities_one_valid))

AUC ROC: 0.8505866226089733


Конечно не лучший AUC-ROC за исследование, но для лучшей модели из всех довольно хорошо, теперь проверим матрицу ошибок:

In [ ]:
cm = confusion_matrix(target_valid, pred_balanced_rfc)

fig = go.Figure(data=go.Heatmap(
        z=cm,
        x=['Предсказала 0', 'Предсказала 1'],
        y=['Правильно 0', 'Правильно 1'],
        colorscale='sunsetdark'
    ))

fig.update_layout(
    xaxis_title="Predicted",
    yaxis_title="True"
)

fig.show()


Неплохой результат, после upsampling'а видно, что модель чаще идет предсказывать положительный класс, чем было до этого, так что увеличение выборки справилось со своей работой. Протестируем downsampling.

Увеличенный диапазон увеличил AUC и уменьшил f1-меру. В итоге получаем, что с увеличением диапазона гиперпараметров увеличивается AUC-ROC и уменьшается f1-мера.

### Downsampling

Хоть мы и решили поставленную задачу, давайте проверим сможет ли лучший наш алгоритм машинного обучения случайный лес добиться 0.6 по f1-мере на тестовой выборке, если мы уменьшим выборку для баланса классов.

In [ ]:
def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones])
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])
    
    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=12345)
    
    return features_downsampled, target_downsampled

In [ ]:
features_downsampled, target_downsampled = downsample(features_train, target_train, 0.4)


В целях экономии времени следующую ячейку можно не запускать.

In [ ]:
%%time
space = {
    'criterion': hp.choice('criterion', ['gini', 'entropy']),
    'max_features': hp.choice('max_features', [None, 'sqrt', 'log2']),
    'max_depth': hp.choice('max_depth', range(2, 50)),
    'min_samples_leaf': hp.choice('min_samples_leaf', range(2, 20)),
    'min_samples_split': hp.choice('min_samples_split',range(2, 20)),
    'n_estimators': hp.choice('n_estimators',range(10, 200))
}

def objective(params):
    clf = RandomForestClassifier(
        criterion=params['criterion'],
        max_depth=params['max_depth'],
        min_samples_split=params['min_samples_split'],
        min_samples_leaf=params['min_samples_leaf'],
        max_features=params['max_features'],
        n_estimators = params['n_estimators'],
        random_state=12345
    )

    clf.fit(features_downsampled, target_downsampled)

    y_pred = clf.predict(features_valid)

    score = f1_score(target_valid, y_pred)

    return {'loss': -score, 'params': params, 'status': STATUS_OK}

trials = Trials()

try:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.RandomState(12345))
except:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.default_rng(12345))

print("Best hyperparameters:")
print(best)

downsampled_rfc = RandomForestClassifier(
    max_depth=range(2, 50)[best['max_depth']],
    min_samples_split=range(2, 20)[best['min_samples_split']],
    min_samples_leaf=range(2, 20)[best['min_samples_leaf']],
    criterion=['gini', 'entropy'][best['criterion']],
    max_features=[None, 'sqrt', 'log2'][best['max_features']],
    n_estimators = range(10, 200)[best['n_estimators']],
    random_state=12345
)
downsampled_rfc.fit(features_downsampled, target_downsampled)

100%|██████████| 100/100 [00:48<00:00,  2.06trial/s, best loss: -0.6159999999999999]
Best hyperparameters:
{'criterion': 0, 'max_depth': 30, 'max_features': 2, 'min_samples_leaf': 2, 'min_samples_split': 6, 'n_estimators': 159}
CPU times: user 48.6 s, sys: 286 ms, total: 48.9 s
Wall time: 49.2 s


RandomForestClassifier(max_depth=32, max_features='log2', min_samples_leaf=4,
                       min_samples_split=8, n_estimators=169,
                       random_state=12345)

In [ ]:
downsampled_rfc = RandomForestClassifier(max_depth=39, max_features='sqrt', min_samples_leaf=12,
                       n_estimators=194, random_state=12345)
downsampled_rfc.fit(features_downsampled, target_downsampled)

RandomForestClassifier(max_depth=39, max_features='sqrt', min_samples_leaf=12,
                       n_estimators=194, random_state=12345)

In [ ]:
test_pred = downsampled_rfc.predict(features_test)
print('F1-SCORE:', f1_score(target_test, test_pred))

F1-SCORE: 0.5883807169344869


Добились максимального результата на валидационной выборке, но на тестовой всего 0.58 по f1-мере. Возможно это связано с недообучением, так как часть строк мы просто уничтожили. Подведем итог.

### Вывод
Среди трех различных алгоритмов машинного обучения, случайный лес показал лучший результат и все-таки смог решить нашу задачу. Мы смогли добиться 0.6 показателя f1-меры на тестовой выборке, мы использовали два способа борьбы с дисбалансом классов, взвешивание и upsampling. Взвешивание дало один из лучших результатов на валидационной выборке по показателю f1-меры (0.61) и по показателю площади по ROC-кривой (0.85), в то время как модель случайного леса не смогла решить поставленную задачу и добиться 0.59 на тестовой выборке (всего 0.58). Дальше мы решили использовать увеличение выборки как способ борьбы с дисбалансом классов, получили 0.6 на валидационной выборке и 0.6 на тестовой выборке. В конце мы решили протестировать, что было бы, если бы мы наоборот уменьшили выборку, получили лучший результат по f1-мере (0.61), но поставленную задачу мы не решили.

### Бонус модель градиентного бустинга

Хоть и по курсу у нас еще не было моделей градиентного бустинга и я не совсем понимаю, как она устроена "под капотом", но все равно хотелось бы проверить даст ли она результат лучше, чем случайный лес. По сути градиентный бустинг это последовательное создание деревьев. Результаты предыдущего дерева влияют на следующие, как бы создание сбалансированного дерева, проверим, какой результат нам даст алгоритм градиентного бустинга. Следующую ячейку лучше не запускать.

In [ ]:
%%time
space = {
    'learning_rate': hp.choice('learning_rate', np.arange(0.01, 0.51, 0.01)),
    'max_features': hp.choice('max_features', [None, 'sqrt', 'log2']),
    'max_depth': hp.choice('max_depth', range(2, 20)),
    'min_samples_leaf': hp.choice('min_samples_leaf', range(2, 20)),
    'min_samples_split': hp.choice('min_samples_split',range(2, 20)),
    'n_estimators': hp.choice('n_estimators',range(100, 1100, 50)),
    'subsample': hp.choice('subsample', np.arange(0.5, 1, 0.1))
}

def objective(params):
    clf = GradientBoostingClassifier(
        learning_rate=params['learning_rate'],
        max_depth=params['max_depth'],
        min_samples_split=params['min_samples_split'],
        min_samples_leaf=params['min_samples_leaf'],
        max_features=params['max_features'],
        n_estimators = params['n_estimators'],
        subsample = params['subsample'],
        random_state=12345
    )

    clf.fit(features_upsampled, target_upsampled)

    y_pred = clf.predict(features_valid)

    score = f1_score(target_valid, y_pred)

    return {'loss': -score, 'params': params, 'status': STATUS_OK}

trials = Trials()

try:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.RandomState(12345))
except:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.default_rng(12345))

print("Best hyperparameters:")
print(best)

model_db = GradientBoostingClassifier(
    max_depth=range(2, 50)[best['max_depth']],
    min_samples_split=range(2, 20)[best['min_samples_split']],
    min_samples_leaf=range(2, 20)[best['min_samples_leaf']],
    learning_rate=np.arange(0.01, 0.51, 0.01)[best['learning_rate']],
    max_features=[None, 'sqrt', 'log2'][best['max_features']],
    n_estimators = range(10, 200)[best['n_estimators']],
    subsample = np.arange(0.4, 1, 0.1)[best['subsample']],
    random_state=12345)
model_db.fit(features_upsampled, target_upsampled)

100%|██████████| 100/100 [12:29<00:00,  7.50s/trial, best loss: -0.6047087980173482]
Best hyperparameters:
{'learning_rate': 6, 'max_depth': 3, 'max_features': 2, 'min_samples_leaf': 2, 'min_samples_split': 8, 'n_estimators': 2, 'subsample': 2}
CPU times: user 12min 23s, sys: 1.86 s, total: 12min 25s
Wall time: 12min 29s


GradientBoostingClassifier(learning_rate=0.06999999999999999, max_depth=5,
                           max_features='log2', min_samples_leaf=4,
                           min_samples_split=10, n_estimators=12,
                           random_state=12345, subsample=0.6)

In [ ]:
model_db = GradientBoostingClassifier(learning_rate=0.01, max_depth=6, min_samples_leaf=12,
                           min_samples_split=12, n_estimators=17,
                           random_state=12345, subsample=0.4)
model_db.fit(features_upsampled, target_upsampled)
pred_db = model_db.predict(features_test)
print(f1_score(target_test, pred_db))

0.5770750988142292


Градиентный бустинг из библиотеки sklearn не справился, всего 0.57 по f1-мере, попробуем из библиотеки catboost. Опять же лучше не ждать и пропускать ячейку.

In [ ]:
space = {
    'learning_rate': hp.loguniform('learning_rate', -6, 0),
    'depth': hp.choice('depth', range(1, 10)),
    'l2_leaf_reg': hp.loguniform('l2_leaf_reg', -16, 2),
    'c': hp.choice('border_count', range(32, 256, 32))
}
def objective(params):
    clf = CatBoostClassifier(
        learning_rate=params['learning_rate'],
        depth=params['depth'],
        l2_leaf_reg=params['l2_leaf_reg'],
        random_state=12345, verbose=False
    )

    clf.fit(features_upsampled, target_upsampled)

    y_pred = clf.predict(features_valid)

    score = f1_score(target_valid, y_pred)

    return {'loss': -score, 'params': params, 'status': STATUS_OK}
trials = Trials()
try:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.RandomState(12345))
except:
    best = fmin(
        fn=objective,
        space=space,
        algo=tpe.suggest,
        max_evals=100,
        trials=trials,
        rstate=np.random.default_rng(12345))
print("Best set of hyperparameters:")
print(best)

100%|██████████| 100/100 [06:50<00:00,  4.10s/trial, best loss: -0.6134347275031686]
Best set of hyperparameters:
{'border_count': 3, 'depth': 4, 'l2_leaf_reg': 0.7783864554247678, 'learning_rate': 0.033021152467940226}


In [ ]:
best_gb = CatBoostClassifier(depth=range(1, 10)[best['depth']], border_count=range(32, 256, 32)[best['border_count']], l2_leaf_reg=best['l2_leaf_reg'], learning_rate=best['learning_rate'], verbose=False, random_state=12345)
best_gb.fit(features_upsampled, target_upsampled)
test_pred = best_gb.predict(features_test)
print(f1_score(target_test, test_pred))

0.5983701979045402


Да, а вот и вторая модель, которая смогла пройти порог в 0.59 по f1-мере, перейдем к тестированию.

## Тестирование модели

Поскольку мы уже знаем, две наши модели смогли успешно решить задачу, проверим их на вменяемость.

In [ ]:
dummy = DummyClassifier(strategy='most_frequent')
dummy.fit(features_upsampled, target_upsampled)
dummy_pred = dummy.predict(features_test)
print(f1_score(target_test, dummy_pred))

0.3547511312217194


Абсолютно без шансов против наших двух моделей, обе имеют 0.6 по f1-мере, а теперь посмотрим на площадь по ROC-кривой:

In [ ]:
probabilities_valid = dummy.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_dummy = roc_auc_score(target_valid, probabilities_one_valid)
print(auc_dummy)

0.5


Да, ожидаемо 0.5, хотя наши модели имеют 0.8+, и тут они лучше по адекватности.

## Вывод по проекту
В наше распоряжение попал датасет, содержащий информацию по клиентам банка, главной задачей было сделать модель машинного обучения, которая бы предсказывала уйдет ли клиент из банка или нет. Сначала мы начали готовить данные для моделей машинного обучения, были удалены столбцы, которые бы мешали нашей модели действовать (столбец с фамилиями клиентов или id-клиентов), после мы обнаружили, что у нас остался довольно важный столбец, а именно, в какой стране пользователь проживает, так как стран было мало (всего 3), то мы применили прямое кодирование для того, чтобы логистическая регрессия работала. Дальше стандартизировали масштаб признаков и уже на этом моменте обнаружили явный дисбаланс в данных, было решено провести исследование задачи дважды, до балансировки и после. Ожидаемо, модели, которые обучались на несбалансированных данных показали результат, который оказался хуже, чем у моделей с балансированными данными (случайный лес при дисбалансе добился 0.56 по f1-мере для тестовой выборки и 0.84 по AUC-ROC), после этого было решено исправить дисбаланс в данных. Было применено два способа, а именно взвешивание классов и изменение размера выборки, взвешивание классов показала результат, который все-таки не дотягивал до приемлимого (случайный лес показал 0.58 по f1-мере и 0.85 по AUC-ROC). Дальше в ход пошел upsampling, который как раз-таки смог добиться приемлимого результата (0.61 по f1-мере у случайного леса и  0.84 по ROC-AUC). Ради интереса я проверил показатель при downsampling выборки и он оказался все-таки хуже, чем нужно (0.58 по f1-мере). По мере исследования была выбрана статегия подбора оптимальных гиперпараметров, а именно байесовская оптимизация, которая минимизировала отрицательную метрику f1-меры. Также для эксперимента решил попробовать использовать модель градиентного бустинга из библиотеки catboost и получил результат, который также удовлетворял нашему исследованию. Итого мы подобрали и обучили две модели машинного обучения, которые решили нашу задачу

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: задача исследована
    - [x]  Исследован баланс классов
    - [x]  Изучены модели без учёта дисбаланса
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 3: учтён дисбаланс
    - [x]  Применено несколько способов борьбы с дисбалансом
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 4: проведено тестирование
- [x]  Удалось достичь *F1*-меры не менее 0.59
- [x]  Исследована метрика *AUC-ROC*